### PR TITLE
Adding support to use docker creds in convox build and services pod

### PR DESCRIPTION
**Pull request by numcys**
### What is the feature/fix?
Adding support for using convox registries in convox build and services.

### CODE REVIEW

File: provider/k8s/app.go [hunk 1]
```diff
@@ -52,6 +52,19 @@ func (p *Provider) AppCreate(name string, opts structs.AppCreateOptions) (*struc
 		return nil, errors.WithStack(err)
 	}
 
+	sourceSecret, _ := p.Cluster.CoreV1().Secrets(p.RackName+"-system").Get(context.TODO(), "registries", am.GetOptions{})
```
Comment (nightfury1204) [inline provider/k8s/app.go:55]: why are you ignoring error here? you should not ignore error

also use `p.Namespace` for system namespace

File: provider/k8s/app.go [hunk 2]
```diff
@@ -52,6 +52,19 @@ func (p *Provider) AppCreate(name string, opts structs.AppCreateOptions) (*struc
...
+	if sourceSecret != nil {
+		newSecret := &ac.Secret{
+			ObjectMeta: am.ObjectMeta{
+				Name:      sourceSecret.Name,
+				Namespace: p.AppNamespace(name),
+			},
+			Type: sourceSecret.Type,
+			Data: sourceSecret.Data,
+		}
+		p.Cluster.CoreV1().Secrets(p.AppNamespace(name)).Create(context.TODO(), newSecret, am.CreateOptions{})
```
Comment (nightfury1204) [inline provider/k8s/app.go:65]: why are you ignoring error?

File: terraform/api/k8s/main.tf [hunk 3]
```diff
@@ -106,6 +106,10 @@ resource "kubernetes_deployment" "api" {
           }
         }
 
+        image_pull_secrets {
+          name = "registries"
```
Comment (nightfury1204) [inline terraform/api/k8s/main.tf:110]: why did explicitly specify `registries` here?

**Review by nightfury1204 — CHANGES_REQUESTED** [review]
Hey why you did not choose docker hub authentication here?

File: provider/k8s/process.go [hunk 4]
```diff
@@ -573,6 +575,34 @@ func (p *Provider) podSpecFromRunOptions(app, service string, opts structs.Proce
 		s.Containers[0].Resources.Limits["memory"] = resource.MustParse(fmt.Sprintf("%dMi", *opts.MemoryLimit))
 	}
 
+	p.DockerUsername = common.CoalesceString(os.Getenv("DOCKER_HUB_USERNAME"),"")
+	p.DockerPassword = common.CoalesceString(os.Getenv("DOCKER_HUB_PASSWORD"),"")
```
Comment (nightfury1204) [inline provider/k8s/process.go:579]: do not initialize the var here. you should initialize that in here: https://github.com/convox/convox/blob/master/provider/k8s/k8s.go#L94

File: provider/k8s/process.go [hunk 5]
```diff
@@ -573,6 +575,34 @@ func (p *Provider) podSpecFromRunOptions(app, service string, opts structs.Proce
...
+
+	_, err = p.CreateOrPatchSecret(p.ctx, am.ObjectMeta{
```
Comment (nightfury1204) [inline provider/k8s/process.go:581]: add a check if docker username and password is not provided, don't execute these logic

**Review by nightfury1204 — CHANGES_REQUESTED** [review]
I don't see image pull secrets related changes for atom and resolver here, is it already provided?

**Comment by numcys** [comment]
Yes Nahid, Atom and resolver already are provided.

File: provider/k8s/k8s.go [hunk 6]
```diff
@@ -151,6 +153,8 @@ func FromEnv() (*Provider, error) {
 		ConvoxDomainTLSCertDisable: os.Getenv("CONVOX_DOMAIN_TLS_CERT_DISABLE") == "true",
 		CertManagerClient:          cm,
 		DiscoveryClient:            kc.Discovery(),
+		DockerUsername: 			common.CoalesceString(os.Getenv("DOCKER_HUB_USERNAME"),""),
+		DockerPassword: 			common.CoalesceString(os.Getenv("DOCKER_HUB_PASSWORD"),""),
```
Comment (nightfury1204) [inline provider/k8s/k8s.go:157]: common.CoalesceString is not needed here
currently it is like if the value is empty then set empty.
fix this please

**Review by nightfury1204 — CHANGES_REQUESTED** [review]

File: provider/k8s/k8s.go (hunk 6 above)
Comment (numcys) [reply provider/k8s/k8s.go]: Added the changes...

**Review by nightfury1204 — DISMISSED** [review]

**Pull request merged**